### PR TITLE
Implement pkg/agent/core/ngt/router test

### PR DESCRIPTION
**Pull request by kevindiu**
### Description:

This PR implements pkg/agent/core/ngt/router test.
It also refactors internal/http routers function so that we can get the route by name in test case.
*Author review required

### Environment:

- Go Version: 1.16
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.3

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [x] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [x] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you successfully ran tests with your changes locally?

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by vankichi** [comment]
Could you please explain the purpose of why you refactored?

**Comment by vankichi** [comment]
Could you please explain the purpose of why you refactored?

**Comment by kevindiu** [comment]
@vankichi Why I refactored is because when writing unit test, I want to get the route information (path, handlerFunc, etc) initialized in http.Handler. 
Without using named routes  in  router.go, there is no another way to get the information above, so I refactored to use `.Name(route.Name)`.
Ref of getting named routes: https://github.com/gorilla/mux/blob/master/mux.go#L214-L216

**Comment by vankichi** [comment]
@kevindiu should we create the test about opts is nil?

**Comment by kevindiu** [comment]
@vankichi the panic will occurred if we don't pass opts into it because the handler is nil.
Panic here: https://github.com/vdaas/vald/blob/a34fd2f2b2da039b967e27ab5780d05af73dcaef/pkg/agent/core/ngt/router/router.go#L58

```   
--- FAIL: TestNew/return_handlers_when_opts_is_empty (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x47a01af]

goroutine 10 [running]:
testing.tRunner.func1.2(0x4856220, 0x4f03570)
	/usr/local/Cellar/go/1.16.3/libexec/src/testing/testing.go:1143 +0x332
testing.tRunner.func1(0xc0004a6900)
	/usr/local/Cellar/go/1.16.3/libexec/src/testing/testing.go:1146 +0x4b6
panic(0x4856220, 0x4f03570)
	/usr/local/Cellar/go/1.16.3/libexec/src/runtime/panic.go:965 +0x1b9
github.com/vdaas/vald/pkg/agent/core/ngt/router.New(0x4f4d510, 0x0, 0x0, 0xc00006df08, 0x4cf44112)
	/Users/wdiu/go/src/github.com/vdaas/vald/pkg/agent/core/ngt/router/router.go:58 +0x24f
github.com/vdaas/vald/pkg/agent/core/ngt/router.TestNew.func4(0xc0004a6900)
	/Users/wdiu/go/src/github.com/vdaas/vald/pkg/agent/core/ngt/router/router_test.go:325 +0x177
```

**Review by vankichi — APPROVED** [review]

**Comment by kevindiu** [comment]
@kpango @rinx can you please review this PR when you have time?
I have refactored the implementation so that I can able to implement unit test.

The reason is written here and the PR description.
https://github.com/vdaas/vald/pull/1214#issuecomment-836116290

**Comment by rinx** [comment]
Could you check `timeout` is set to the returned handler correctly?
https://github.com/vdaas/vald/blob/test/pkg/add-agent-core-ngt-router-test/pkg/agent/core/ngt/router/option.go#L37

**Comment by kevindiu** [comment]
@rinx I think it should be test here.
https://github.com/vdaas/vald/pull/1206/files#diff-ea76e110ce4a628ff21967be02fa41ec4116246ffe4f6a21b5d152c2915ef14eL145

**Comment by rinx** [comment]
Sorry, the link is incorrect.
Please check it in here.
https://github.com/vdaas/vald/blob/test/pkg/add-agent-core-ngt-router-test/pkg/agent/core/ngt/router/router_test.go#L48

**Comment by rinx** [comment]
If it is impossible, please ignore it :cat:

File: internal/net/http/routing/router.go [hunk 1]
```diff
@@ -51,7 +51,7 @@ func New(opts ...Option) http.Handler {
 
 		rt.Handle(route.Pattern,
 			r.routing(route.Name, route.Pattern,
-				route.Methods, route.HandlerFunc))
+				route.Methods, route.HandlerFunc)).Name(route.Name)
```
Comment (kpango) [inline internal/net/http/routing/router.go:54]: why do we need name??

Comment (kevindiu) [reply internal/net/http/routing/router.go:54]: @kpango the reason is written here. 
https://github.com/vdaas/vald/pull/1214#issuecomment-836116290

The main reason is to able to get the route reference inside mux in test case

**Review by kpango — APPROVED** [review]
LGTM

**Pull request merged**